Added support for object initializers and member initializers

333fred opened the pull request:
Object initializers are now supported, as are nested member initializers. For member initializers,
I inspected the IL we generate and ensured that we are duplicating accesses as appropriate.

I additionally addressed comments from the previous PR, https://github.com/dotnet/roslyn/pull/25322.
@dotnet/roslyn-compiler @dotnet/analyzer-ioperation @AlekseyTs for reviews please.

333fred commented:
@AlekseyTs looking at test failures.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3104,13 +3106,97 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+                AddStatement(new FlowCapture(initializerCaptureId, objectCreation.Syntax, objectCreation));
 
-            // PROTOTYPE(dataflow): For a non-null initializer, we'll need to assign the created object to a flow reference and rewrite all initializers to assign to/call functions on it, then return return the flow reference
+                objectCreation = new FlowCaptureReference(initializerCaptureId, objectCreation.Syntax, objectCreation.Type, objectCreation.ConstantValue);
```
>objectCreation [](start = 16, length = 14)

It feels like the name ```objectCreation``` is somewhat confusing after this point. It is no longer an object creation. Consider renaming the local to "createdInstance" or something similar.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3104,13 +3106,97 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+            // We therefore visit the InitializedMember to get the implicit receiver for the contained initializer, and that implicit receiver will be cloned everywhere it encounters
+            // an IInstanceReferenceOperation with ReferenceKind InstanceReferenceKind.ImplicitReceiver
+
+            IOperation implicitReceiver = Visit(operation.InitializedMember);
```
>implicitReceiver [](start = 23, length = 16)

The name is feels misleading, it looks like this is not a receiver, but an instance to initialize. Consider renaming.

AlekseyTs replied:
After looking at what HandleObjectOrCollectionInitializer is doing, the name doesn't sound that surprising any more, but it is still feels like this terminology is specific to internal workings of HandleObjectOrCollectionInitializer 

---
In reply to: [184474823](https://github.com/dotnet/roslyn/pull/26389#discussion_r184474823) [](ancestors = 184474823)

AlekseyTs commented:
>         public override IOperation VisitObjectOrCollectionInitializer(IObjectOrCollectionInitializerOperation operation, int? captureIdForResult)

Should we throw from this method?

---
Refers to: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:3513 in 65a51d8. [](commit_id = 65a51d8e148ef4d510645f08e790d897d745b24e, deletion_comment = False)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3104,13 +3106,97 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+            return null;
+        }
+
+        private void HandleObjectOrCollectionInitializer(IObjectOrCollectionInitializerOperation initializer, IOperation implicitReceiver)
```
>implicitReceiver [](start = 121, length = 16)

Consider renaming to "instanceToInitialize" or something similar.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3104,13 +3106,97 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+                // When we're in an object or collection initializer, we need to replace the instance reference with a reference to the object being initialized
+                Debug.Assert(_currentImplicitReceiver != null);
+                Debug.Assert(operation.IsImplicit);
+                return OperationCloner.CloneOperation(_currentImplicitReceiver);
```
>_currentImplicitReceiver [](start = 54, length = 24)

It looks like this could be an arbitrary expression that might include nodes that are only created by this builder. Currently, OperationCloner throws on many of them. I think we need to implement cloning for them after all.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs:
```diff
@@ -1142,5 +1142,1786 @@ static void M(bool b)
+              Right: 
+                ILiteralOperation (OperationKind.Literal, Type: System.Int32, Constant: 1) (Syntax: '1')
+
+            IFlowCaptureOperation: 1 (OperationKind.FlowCapture, Type: null, IsImplicit) (Syntax: 'B')
```
>IFlowCaptureOperation: 1 (OperationKind.FlowCapture, Type: null, IsImplicit) (Syntax: 'B') [](start = 12, length = 90)

It looks like this capture can be avoided.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs:
```diff
@@ -1142,5 +1142,1786 @@ static void M(bool b)
+              Right: 
+                ILiteralOperation (OperationKind.Literal, Type: System.Int32, Constant: 3) (Syntax: '3')
+
+            IFlowCaptureOperation: 1 (OperationKind.FlowCapture, Type: null, IsImplicit) (Syntax: 'P2')
```
>IFlowCaptureOperation: 1 (OperationKind.FlowCapture, Type: null, IsImplicit) (Syntax: 'P2') [](start = 12, length = 91)

It feels like this capture can be limited to capturing just the instance receiver

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs:
```diff
@@ -1142,5 +1142,1786 @@ static void M(bool b)
+                            IFlowCaptureReferenceOperation: 0 (OperationKind.FlowCaptureReference, Type: C1, IsImplicit) (Syntax: 'new C1 { C2 ... ] = 4 } } }')
+                  Arguments(1):
+                      IArgumentOperation (ArgumentKind.Explicit, Matching Parameter: i) (OperationKind.Argument, Type: null) (Syntax: '1')
+                        ILiteralOperation (OperationKind.Literal, Type: System.Int32, Constant: 1) (Syntax: '1')
```
>ILiteralOperation (OperationKind.Literal, Type: System.Int32, Constant: 1) (Syntax: '1') [](start = 24, length = 88)

It looks like the order of evaluation doesn't match what the compiler emits. You might want to use a function call here to make this clearly observable. We might also want to adjust ```Children``` property behavior to reflect the actual order of evaluation.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs:
```diff
@@ -1142,5 +1142,1786 @@ static void M(bool b)
+
+    public void M(bool b)
+    /*<bind>*/{
+        var x = new C1 { C2 = { C31 = { [1] = 1, [2] = 2 }, C32 = { [3] = 3, [4] = 4 } } };
```
>1 [](start = 46, length = 1)

If you are using literals, consider avoiding using the same value multiple times, that will make it easier to make sense of the base-line.

AlekseyTs replied:
>1 [](start = 41, length = 1)

It would be good to have tests covering more than one index.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs:
```diff
@@ -1142,5 +1142,1786 @@ static void M(bool b)
+                      IArgumentOperation (ArgumentKind.Explicit, Matching Parameter: i) (OperationKind.Argument, Type: null) (Syntax: '1')
+                        ILiteralOperation (OperationKind.Literal, Type: System.Int32, Constant: 1) (Syntax: '1')
+                        InConversion: CommonConversion (Exists: True, IsIdentity: True, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
+                        OutConversion: CommonConversion (Exists: True, IsIdentity: True, IsNumeric: False, IsReference: False, IsUserDefined: False) (Method
```
>IFlowCaptureReferenceOperation: 1 (OperationKind.FlowCaptureReference, Type: System.Int32, IsImplicit) (Syntax: 'b ? 4 : 5') [](start = 24, length = 124)

Here the order of evaluation looks correct, but it looks like this is an accident because we haven't yet implemented proper handling in VisitPropertyReference. Given current implementation, object initializers depend on proper implementation for property references. Consider either adding proper implementation to VisitPropertyReference in context of this PR, or submitting a separate PR, merging it first and then merging it in this PR.

AlekseyTs replied:
>} [](start = 8, length = 1)

It would be good to have some tests covering fields initialization. Some existing tests can simply be changed to target fields

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3104,13 +3106,97 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+
+            foreach (IOperation innerInitializer in initializer.Initializers)
+            {
+                AddStatement(Visit(innerInitializer));
```
>Visit(innerInitializer) [](start = 29, length = 23)

Rather than simply relying on default visit behavior, we probably should explicitly handle the initializers that we expect to run into, at least for success scenarios, and leave Visit for things we don't recognize.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs:
```diff
@@ -1142,5 +1142,1786 @@ static void M(bool b)
+
+    public void M()
+    /*<bind>*/{
+        var x = new C1 { C2 = { C31 = { P1 = 1, P2 = 2 } } };
```
>C31 [](start = 32, length = 3)

It would be good to have a test with an indexer in this position, including coalesce expression as an index. We should also ensure the evaluation order is correct.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs:
```diff
@@ -1142,5 +1142,1786 @@ static void M(bool b)
+
+    public void M(bool b)
+    /*<bind>*/{
+        var x = new C1 { C2 = { C31 = { P1 = 1, P2 = 2 }, C32 = b ? ({ P1 = 3, P2 = 4 }) : ({ P1 = 3, P2 = 4 })
```
>C2 [](start = 25, length = 2)

Is it possible to get something other than field/property reference on the left hand side of an initializer? It would be good to have a targeted test for that

333fred replied:
Events are the only other thing, and I added tests for that. I have a test asserting that you can't treat the lhs as an array (and access elements). The LocalRewriter throws on anything other than those for object initializers.

---
In reply to: [184517021](https://github.com/dotnet/roslyn/pull/26389#discussion_r184517021) [](ancestors = 184517021)

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.vb:
```diff
@@ -872,6 +872,769 @@ Block[B5] - Exit
+                  Operand: 
+                    ILiteralOperation (OperationKind.Literal, Type: System.Int32, Constant: 1) (Syntax: '1')
+
+            IFlowCaptureOperation: 1 (OperationKind.FlowCapture, Type: null, IsImplicit) (Syntax: 'B')
```
>IFlowCaptureOperation: 1 (OperationKind.FlowCapture, Type: null, IsImplicit) (Syntax: 'B') [](start = 12, length = 90)

It looks like this capture can be avoided.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.vb:
```diff
@@ -872,6 +872,769 @@ Block[B5] - Exit
+                  Operand: 
+                    ILiteralOperation (OperationKind.Literal, Type: System.Int32, Constant: 1) (Syntax: '1')
+
+            IFlowCaptureOperation: 3 (OperationKind.FlowCapture, Type: null, IsImplicit) (Syntax: 'B')
```
>IFlowCaptureOperation: 3 (OperationKind.FlowCapture, Type: null, IsImplicit) (Syntax: 'B') [](start = 12, length = 90)

It looks like this capture can be avoided

AlekseyTs commented:
Done with review pass (iteration 2)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs:
```diff
@@ -1142,5 +1142,1786 @@ static void M(bool b)
+
+    public void M(bool b)
+    /*<bind>*/{
+        var x = new C1 { C2 = { C31 = { [1] = 1, [2] = 2 }, C32 = { [3] = 3, [4] = 4 } } };
```
>[2] [](start = 49, length = 3)

It would also be interesting to cover indexing into an array returned by a property and by a field

AlekseyTs replied:
Also,  indexing into an object with an indexer returned by a property and by a field  

---
In reply to: [184542172](https://github.com/dotnet/roslyn/pull/26389#discussion_r184542172) [](ancestors = 184542172)

333fred replied:
> It would also be interesting to cover indexing into an array returned by a property and by a field 
I'll add a prototype comment to add support for this. For this PR, I'm just going to support fields, events, and properties. I'll add array support at a later time, likely when I add support for array initializers as well.

---
In reply to: [184542172](https://github.com/dotnet/roslyn/pull/26389#discussion_r184542172) [](ancestors = 184542172)

333fred replied:
> Also,  indexing into an object with an indexer returned by a property and by a field  
I'll have this, as well as nested objects that have initializers.

---
In reply to: [185285865](https://github.com/dotnet/roslyn/pull/26389#discussion_r185285865) [](ancestors = 185285865,184542172)

cston commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3104,11 +3108,215 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+                            propertyArguments = propertyArgumentsBuilder.ToImmutableAndFree();
+                        }
+
+                        Debug.Assert((propertyReference == null && propertyArguments == ImmutableArray<IArgumentOperation>.Empty) ||
```
>propertyArguments == ImmutableArray<IArgumentOperation>.Empty [](start = 67, length = 61)

`propertyArguments.IsEmpty`

cston commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3104,11 +3108,215 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+
+            void handleMemberInitializer(IMemberInitializerOperation memberInitializer)
+            {
+                // We explicitly do not push the initialized member onto the stack here. We visit the initialized member to get the implicit receiver that will be substitued in when an
```
>substitued [](start = 163, length = 10)

Typo

cston commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3104,11 +3108,215 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+            void handleMemberInitializer(IMemberInitializerOperation memberInitializer)
+            {
+                // We explicitly do not push the initialized member onto the stack here. We visit the initialized member to get the implicit receiver that will be substitued in when an
+                // IInstanceReferenceOperation with InstanceReferenceKind.ImplicitReceiver is encounted. If that receiver needs to be pushed onto the stack, its parent will handle it.
```
>encounted [](start = 94, length = 9)

Typo

cston commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3104,11 +3108,215 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+            {
+                // We explicitly do not push the initialized member onto the stack here. We visit the initialized member to get the implicit receiver that will be substitued in when an
+                // IInstanceReferenceOperation with InstanceReferenceKind.ImplicitReceiver is encounted. If that receiver needs to be pushed onto the stack, its parent will handle it.
+                // In member initializers, the code generated will evalute InitializedMember multiple times. For example, if you have the following:
```
>evalute [](start = 67, length = 7)

Typo

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [384782136](https://github.com/dotnet/roslyn/pull/26389#issuecomment-384782136) [](ancestors = 384782136)

333fred commented:
@dotnet-bot test this please

333fred commented:
@dotnet-bot test this please

mmitche commented:
@dotnet-bot test this please

mmitche commented:
@dotnet-bot test this please

jaredpar commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -32,6 +32,11 @@ internal sealed partial class ControlFlowGraphBuilder : OperationVisitor<int?, I
+        /// <summary>
+        /// Holds the current object being initialized if we're visiting an object initializer.
+        /// </summary>
+        private IOperation _currentInitializedInstance = null;
```
Why the explicit initialization here? It's unneeded.

jaredpar commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -1029,6 +1034,7 @@ private void VisitStatement(IOperation operation)
             IOperation saveCurrentStatement = _currentStatement;
             _currentStatement = operation;
             Debug.Assert(_evalStack.Count == 0);
+            Debug.Assert(_currentConditionalAccessInstance == null);
```
Did you mean `_currentInitializedInstance` here?

jaredpar commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3104,11 +3108,215 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+                        if (memberReference.Kind == OperationKind.PropertyReference &&
+                            !(propertyReference = (IPropertyReferenceOperation)memberReference).Arguments.IsEmpty)
+                        {
+                            var propertyArgumentsBuilder = ArrayBuilder<IArgumentOperation>.GetInstance();
```
>.GetInstance(); [](start = 91, length = 15)

Can add propertyReference.Arguments.Count as a parameter to allocate a more efficient builder.

cston replied:
Consider extracting an instance method to visit the argument and calculate `propertyArguments` with `SelectAsArray`. Then the `ArrayBuilder` can be removed, along with the check for `((IPropertyReferenceOperation)memberReference).Arguments.IsEmpty`.
```
var propertyArguments = propertyReference.Arguments.SelectAsArray(
    (arg, visitor) => visitor.HandleSimpleAssignmentArgument(arg),
    this);
```

cston replied:
Never mind, the visit method has side effects so `SelectAsArray` does not feel appropriate.

jaredpar commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3104,11 +3108,215 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+                                left = new PropertyReferenceExpression(propertyReference.Property, instance, propertyArguments, semanticModel: null, memberReference.Syntax,
+                                                                       memberReference.Type, memberReference.ConstantValue, memberReference.IsImplicit);
+                                break;
+                        }
```
>} [](start = 24, length = 1)

Do we need a `default` case here?

333fred replied:
No, we're already inside a switch restricting the kind to these three choices.

---
In reply to: [185550604](https://github.com/dotnet/roslyn/pull/26389#discussion_r185550604) [](ancestors = 185550604)

333fred commented:
Ping @AlekseyTs for an updated review.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3104,11 +3108,215 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+                handleInitializer(innerInitializer);
+            }
+
+            _currentInitializedInstance = previousInitializedInstance;
```
>_currentInitializedInstance = previousInitializedInstance; [](start = 12, length = 58)

If this is the end of the code in this function, consider adding an explicit return statement after it, this would help readability.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3104,11 +3108,215 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+                        throw new NotImplementedException();
+
+                    case OperationKind.Invalid:
+                        AddStatement(Visit(innerInitializer));
```
>AddStatement(Visit(innerInitializer)); [](start = 24, length = 38)

Do we have a test covering a control flow inside ```innerInitializer``` for this case?

333fred replied:
I'm adding one, will be ObjectCreationFlow_37

---
In reply to: [185664097](https://github.com/dotnet/roslyn/pull/26389#discussion_r185664097) [](ancestors = 185664097)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3104,11 +3108,215 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+                        if (memberReference.Kind == OperationKind.PropertyReference &&
+                            !(propertyReference = (IPropertyReferenceOperation)memberReference).Arguments.IsEmpty)
+                        {
+                            var propertyArgumentsBuilder = ArrayBuilder<IArgumentOperation>.GetInstance();
```
>var propertyArgumentsBuilder = ArrayBuilder<IArgumentOperation>.GetInstance(); [](start = 28, length = 78)

Is this a code path for initializer through an indexer?

333fred replied:
It will be.

---
In reply to: [185664894](https://github.com/dotnet/roslyn/pull/26389#discussion_r185664894) [](ancestors = 185664894)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3104,11 +3108,215 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+                                // We assume all arguments have side effects and spill them. We only avoid capturing literals, and
+                                // recapturing things that have already been captured once.
+                                IOperation visitedInnerArgument = Visit(arg.Value);
+                                if (visitedInnerArgument.Kind != OperationKind.Literal &&
```
>visitedInnerArgument.Kind != OperationKind.Literal && [](start = 36, length = 53)

I wouldn't special case literals here. They don't have any special meaning in context of this node, as opposed to string literals in context of IInterpolatedStringTextOperation for example. Also, consider using VisitAndCapture logic, it looks like the code here is pretty much a duplication of what it is doing.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3104,11 +3108,215 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+                // We therefore visit the InitializedMember to get the implicit receiver for the contained initializer, and that implicit receiver will be cloned everywhere it encounters
+                // an IInstanceReferenceOperation with ReferenceKind InstanceReferenceKind.ImplicitReceiver
+
+                IOperation instance = Visit(memberInitializer.InitializedMember);
```
>IOperation instance = Visit(memberInitializer.InitializedMember); [](start = 16, length = 65)

Either I am missing something, or it looks like there is a problem here with the order of evaluation when initialized member is an indexer.

AlekseyTs replied:
>IOperation instance = Visit(memberInitializer.InitializedMember); [](start = 16, length = 65)

Do we have a test covering a control flow inside ```memberInitializer.InitializedMember``` at this point? Inside an index and elsewhere perhaps in error cases. It would be especially interesting to test with coalesce expression.

333fred commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3104,11 +3108,215 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+                    case OperationKind.Invalid:
+                        // If there's an invalid on the left side, we can't make any guarantees about the inner shape, so we don't
+                        // try to handle it specially and fall back to standard handling.
+                        return Visit(assignmentOperation);
```
Yes. ObjectCreationFlow_34 is one example, and I'm adding another.

---
In reply to: [185668191](https://github.com/dotnet/roslyn/pull/26389#discussion_r185668191) [](ancestors = 185668191)

jaredpar commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -1029,6 +1034,7 @@ private void VisitStatement(IOperation operation)
             IOperation saveCurrentStatement = _currentStatement;
             _currentStatement = operation;
             Debug.Assert(_evalStack.Count == 0);
+            Debug.Assert(_currentInitializedInstance == null);
```
> Debug.Assert(_currentInitializedInstance == null); [](start = 11, length = 51)

What happens if the object initializer is a lambda expression? I assume that will cause us to come back to VisitStatement.

333fred replied:
No. While support isn't yet implemented, our decision for lambdas and local functions is that they will be separate graphs, and not put into the main graph. Once you're in an expression, nothing should cause VisitStatement to be called again until you exit that expression.

---
In reply to: [185932789](https://github.com/dotnet/roslyn/pull/26389#discussion_r185932789) [](ancestors = 185932789)

jaredpar commented:
>         [CompilerTrait(CompilerFeature.IOperation, CompilerFeature.Dataflow)]

Duplicate attribute

---
Refers to: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs:4981 in e37a2c4. [](commit_id = e37a2c44ae3171d56486ea6be61a006fcb7b3461, deletion_comment = False)

jaredpar commented:
>         }

Consider two additional tests:

- Initializer includes a lambda with at least one statement 
- Initializer includes a lambda which has a statement that uses an initializer. 

---
Refers to: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs:5781 in e37a2c4. [](commit_id = e37a2c44ae3171d56486ea6be61a006fcb7b3461, deletion_comment = False)

333fred commented:
>         }

I'll add a prototype comment to consider this when lambda support is added. However, I don't think these are interesting scenarios. Lambda bodies do not get visited by this visitor directly, they will only be visited once the lambda has been requested, and at that point they will be in an entirely separate graph that has no relation to the current one.

---
In reply to: [386436089](https://github.com/dotnet/roslyn/pull/26389#issuecomment-386436089) [](ancestors = 386436089)

---
Refers to: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs:5781 in e37a2c4. [](commit_id = e37a2c44ae3171d56486ea6be61a006fcb7b3461, deletion_comment = False)

AlekseyTs commented:
>         }

I think there are interesting scenarios in VB, it might be possible to refer to the instance being initialized inside a lambda. It would be good to make a note of that in the Visit method for lambdas, so that when I get to it, it will be on the radar from the start

---
In reply to: [386455994](https://github.com/dotnet/roslyn/pull/26389#issuecomment-386455994) [](ancestors = 386455994,386436089)

---
Refers to: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs:5781 in e37a2c4. [](commit_id = e37a2c44ae3171d56486ea6be61a006fcb7b3461, deletion_comment = False)

AlekseyTs commented:
>         }

It looks like there is an extra empty line after this function (I see two).

---
Refers to: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs:5781 in e37a2c4. [](commit_id = e37a2c44ae3171d56486ea6be61a006fcb7b3461, deletion_comment = False)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3104,11 +3108,233 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+            }
+
+            IOperation popTarget(IOperation originalTarget, ImmutableArray<IArgumentOperation> arguments)
+            {
```
>{ [](start = 12, length = 1)

Consider removing an empty line below.

333fred commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3104,11 +3108,233 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+                // an IInstanceReferenceOperation with ReferenceKind InstanceReferenceKind.ImplicitReceiver
+
+                (bool pushSuccess, ImmutableArray<IArgumentOperation> arguments) = tryPushTarget(memberInitializer.InitializedMember);
+                IOperation instance = pushSuccess ? popTarget(memberInitializer.InitializedMember, arguments) : Visit(memberInitializer.InitializedMember);
```
No, adding one. This is hit when you have a nested initializer for a property or field that doesn't exist. 

---
In reply to: [185960505](https://github.com/dotnet/roslyn/pull/26389#discussion_r185960505) [](ancestors = 185960505)

AlekseyTs replied:
Could you please make sure we cover control flow in ```memberInitializer.InitializedMember``` when we hit this code path

---
In reply to: [185963830](https://github.com/dotnet/roslyn/pull/26389#discussion_r185963830) [](ancestors = 185963830,185960505)

333fred replied:
Will do. I'll add this in a follow-up PR.

---
In reply to: [185965624](https://github.com/dotnet/roslyn/pull/26389#discussion_r185965624) [](ancestors = 185965624,185963830,185960505)

AlekseyTs replied:
> Will do. I'll add this in a follow-up PR.

Sounds good

---
In reply to: [185966703](https://github.com/dotnet/roslyn/pull/26389#discussion_r185966703) [](ancestors = 185966703,185965624,185963830,185960505)

333fred commented:
@dotnet/analyzer-ioperation @mavasani can I get a review please.

333fred commented:
@dotnet/analyzer-ioperation @mavasani please review

333fred commented:
@mavasani indicated offline that he has no concerns that will not be addressable by follow up PRs, so I will squash and rebase this PR and address any comments in follow ups.
